client/asset/eth/eth.go

```diff
@@ -2371,6 +2462,122 @@ func (eth *ETHWallet) checkForNewBlocks(reportErr func(error)) {
 			w.checkFindRedemptions()
 		}
 	}()
+	go func() {
+		for _, w := range eth.walletList() {
+			w.checkConfirmRedemptions()
+		}
+	}()
+}
+
+func (w *assetWallet) swapIsRedeemed(secretHash [32]byte, contractVersion uint32) (bool, error) {
+	swap, err := w.swap(w.ctx, secretHash, contractVersion)
+	if err != nil {
+		return false, err
+	}
+	return swap.State == dexeth.SSRedeemed, nil
+}
+
+// confirmRedemption checks one of the queued redemption requests.
+func (w *assetWallet) confirmRedemption(txHash common.Hash, req *confirmRedemptionRequest, currentTip int64) {
+	const (
+		TxConfsNeededToConfirm               = 10
+		BlocksToWaitBeforeCoinNotFound       = 5
+		BlocksToWaitBeforeCheckingIfReplaced = 10
+	)
+
+	var blocksSinceRequestStart int64
+	if currentTip > req.startHeight {
+		blocksSinceRequestStart = currentTip - req.startHeight
+	}
+
+	tx, txBlock, err := w.node.getTransaction(w.ctx, txHash)
+	if err == asset.CoinNotFoundError {
+		w.log.Errorf("confirmRedemption for tx %s: could not find coin", txHash)
+		// Wait a few blocks before returning that the coin cannot be found.
+		if blocksSinceRequestStart < BlocksToWaitBeforeCoinNotFound {
+			return
+		}
+	}
+	if err != nil {
+		// If we cannot find a transaction, check if the swap has been redeemed
+		// by another transaction. If so, we can end the search.
+		redeemed, redeemedErr := w.swapIsRedeemed(req.secretHash, req.contractVersion)
+		if redeemedErr != nil {
+			// Just log this error, and return the error from getTransaction.
+			w.log.Errorf("unable to determine if swap has been redeemed: %v", err)
+		}
+		if redeemedErr != nil || !redeemed {
+			req.res <- err
+		} else {
+			// The swap was redeemed by another transaction.
+			req.res <- nil
+		}
+		return
+	}
+
+	var confirmations int64
+	if txBlock > 0 {
+		if txBlock > currentTip {
+			confirmations = 1
+		} else {
+			confirmations = currentTip - txBlock + 1
+		}
+	}
+
+	if confirmations >= TxConfsNeededToConfirm {
+		req.res <- nil
+		return
+	} else if confirmations > 0 {
+		return
+	}
+
+	// If the transaction is still unconfirmed, but was submitted less than 10
+	// blocks ago, give it some time before checking if it has been replaced,
+	// or redeemed by another tx.
+	if blocksSinceRequestStart < BlocksToWaitBeforeCheckingIfReplaced {
+		return
+	}
+
+	redeemed, err := w.swapIsRedeemed(req.secretHash, req.contractVersion)
+	if err != nil {
+		req.res <- err
+		return
+	}
+	if redeemed {
+		req.res <- nil
+		return
+	}
+
+	confirmedNonce, err := w.node.getConfirmedNonce(w.ctx, currentTip)
+	if err != nil {
+		req.res <- err
+		return
+	}
+	if confirmedNonce > tx.Nonce() {
+		req.res <- asset.ErrRedemptionOverridden
+	}
+
+	// Unconfirmed for > 10 blocks, but has not been replaced or redeemed
+	// by another transaction. Need to resubmit with a higher fee. Should
+	// be done is a separate acceleration request, or automatically if
+	// there are enough additional funds?
```

Comment:
> Is this necessarily true? 10 blocks is only a couple of minutes. Shouldn't we check `tx.GasFeeCap() < tipHeader.BaseFee` at least.
> 
> >

Yeah we should check that. If this happens then it means that the base fee is higher than the server's max fee rate which is definitely not good. Do you think it's OK to dip into additional funds that were not reserved for this order in the first place?

> Good question. I almost wish we had a way for wallets to issue a prompt or a `Notification` to the user, but the logistics of that would be crazy, I think.

Why would it be so crazy? I think it's a good idea.

Reply:
> I'm thinking maybe that's the only condition under which we should consider automatically re-broadcasting.

Do you mean in addition the the transaction getting replaced/not found?

The reason I'm hesitant about it is because the user hadn't agreed to spend that much on fees and MaxFee that was returned before the order would have been a lie.

Reply:
What do you think about this: https://github.com/decred/dcrdex/commit/de9fc0e147567504e55e623008cfc5cbed8d1b61

Still untested

Reply:
> What do you think about this: [de9fc0e](https://github.com/decred/dcrdex/commit/de9fc0e147567504e55e623008cfc5cbed8d1b61)
> 
> Still untested

I think that's the basic idea, yeah.